pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java

```diff
@@ -2,19 +2,123 @@
+
+        messageBuilder.sendAsync().whenComplete((producerResult, throwable) -> {
+            if (producerResult != null) {
+                produceFuture.complete(new PulsarProducerResult(producerResult));
+            } else {
+                // In general Pulsar client and producer, auto-reconnects so this should be fine.Might need to
+                // refresh/re-create producer (and possibly client) if there are fatal errors, currently these
+                // failures are unknown.
+                produceFuture.completeExceptionally(throwable);
+            }
+        });
+        return produceFuture;
+    }
+
+
+    private org.apache.pulsar.client.api.Producer<byte[]> getProducer(
+            PulsarStorageTopic topic, ProducerOptions options, String hostname
+    )
+            throws PulsarClientException {
+        Map<String, Object> producerConfig = getProducerConfig(topic, options, hostname);
+        return clientProvider.getPulsarClient().newProducer().loadConf(producerConfig).create();
+    }
+
+    private Map<String, Object> getProducerConfig(PulsarStorageTopic topic, ProducerOptions options, String hostName) {
+
+        // System Configured::
+        // sendTimeoutMs
+        // blockIfQueueFull
+        // compressionType
+        // batchingEnabled
+        // batchingMaxPublishDelay
+
+        // Fixed::
+        // name = topic + hostname
+        // topic = topic name
+        // access mode -- shared.
+        //
+        // Default Values::
+        // enableChunking: false
+        // autoUpdatePartitions, autoUpdatePartitionsInterval
+        //
+        // Calculated:: Topic capacity dependent, (primitive calculation by Varadhi).
+        // maxPendingMessages, maxPendingMessagesAcrossPartitions
+        // batchingMaxPublishDelay, batchingMaxMessages, batchingMaxBytes,
+
+        Map<String, Object> producerConfig = options.asConfigMap();
+        producerConfig.put("topicName", topic.getName());
+        producerConfig.put("producerName", getProducerName(topic.getName(), hostName));
+        producerConfig.put("accessMode", ProducerAccessMode.Shared);
+
+        int topicMaxQps = topic.getMaxQPS();
+        int maxPendingMessages = getMaxPendingMessages(topicMaxQps);
+        int batchingMaxMessages = getBatchMaxMessages(topicMaxQps, options.getBatchingMaxPublishDelayMs());
+        producerConfig.put("maxPendingMessages", maxPendingMessages);
+        // maxPendingMessages and maxPendingMessagesAcrossPartitions are kept same assuming worst case.
+        producerConfig.put("maxPendingMessagesAcrossPartitions", maxPendingMessages);
+        producerConfig.put("batchingMaxMessages", batchingMaxMessages);
+        producerConfig.put("batchingMaxBytes", batchingMaxMessages * getBatchingMaxBytes(batchingMaxMessages, topic));
+        return producerConfig;
     }
+
+    private String getProducerName(String topicName, String hostName) {
+        String topicSuffix = topicName.split("/")[4];
```

Comment:
why split?

Reply:
can be left as is for now. We will evaluate if we need to shorten it later.